core/src/slot_toggled_field.rs

```diff
@@ -24,25 +32,62 @@ impl SlotToggle {
         self.slot_removed
     }
 
-    pub fn activate(&mut self, slot: u64) -> bool {
-        if self.slot_added >= self.slot_removed {
-            false
-        } else {
-            self.slot_added = slot;
-            true
+    pub fn activate(&mut self, slot: u64, epoch_length: u64) -> bool {
+        match self.state(slot, epoch_length) {
+            SlotToggleState::Inactive => {
+                self.slot_added = slot;
+                true
+            }
+            _ => false,
         }
     }
 
-    pub fn deactivate(&mut self, slot: u64) -> bool {
-        if self.slot_added < self.slot_removed {
-            false
-        } else {
-            self.slot_removed = slot;
-            true
+    pub fn deactivate(&mut self, slot: u64, epoch_length: u64) -> bool {
+        match self.state(slot, epoch_length) {
+            SlotToggleState::Active => {
+                self.slot_removed = slot;
+                true
+            }
+            _ => false,
         }
     }
 
-    pub const fn is_active(&self, slot: u64) -> bool {
-        self.slot_added >= self.slot_removed && slot >= self.slot_added
+    pub fn is_active(&self, slot: u64, epoch_length: u64) -> bool {
```

Comment:
perhaps is_staked_or_slashable? is_active conflicts with state. open to other ideas